Add Core Subtopologies: CDHCore

moisesmata opened the pull request:
## Change Description

Within fprime/Svc there is now a Subtopologies folder. The folder contains a subtopolgy, CDHCore, which contains 7 Svc components:
- CommandDispatcher
- ActiveLogger
- Health
- Version
- PassiveTextLogger
- AssetFatalAdapter
- TlmPacketizer/TlmChan

The subtopology's header "PingEntries.hpp" also now contains the PingEntries for each of the relevant components. If desired, users may swap this header for another in their deployment to customize values and have one list for all component PingEntries. 

Ref has been modified to import and use this CDHCore subtopology within the Ref deployment topology. 

Ref integration test has swapped names to refer to CDHCore.cmdDisp and CDHCore.events, instead of Ref.cmdDisp and Ref.eventLogger. 

## Future Work

3 further subtopologies that contain core flight software stacks (Comms, FileHandling, DataProducts) will be added in future pull requests.

moisesmata committed:
Added led-blinker workflow for aarch64-linux platform

moisesmata committed:
Update runner label

moisesmata committed:
Added Tags to RPI self hosted workflows, sparse-checkout-cone-mode false

moisesmata committed:
Add subtopologies to Svc/ restructing of Ref to include CDH subtopology

moisesmata committed:
Merge branch 'nasa:devel' into add-subtopologies

moisesmata committed:
CDHCore Subtopology added under Svc and integrated into Ref

moisesmata committed:
Merge branch 'add-subtopologies' of https://github.com/moisesmata/fprime into add-subtopologies

moisesmata committed:
Health instance within CDHCore references CDHCoreConfig for queue size

moisesmata committed:
Update metadata

check-spelling run (push) for add-subtopologies

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

moisesmata committed:
Added AssertFatalAdapter Version PassiveTextLogger to Subtopology, config clarity improvements

moisesmata committed:
Comms subtopology without cmdSeq created, integrated in Ref

moisesmata committed:
Added cmdSeq to Comms Subtopology, integrated into Ref

moisesmata committed:
Add FileHandling Subtopology, initial structure

moisesmata committed:
Fixed Spelling, Flie -> File

moisesmata committed:
Update metadata

check-spelling run (push) for add-subtopologies

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

moisesmata committed:
Added working FileHandling, DataProducts subtopology, integrated into Ref

moisesmata committed:
prmDb part of FileHandling, fatalHandler part of CDHCore

moisesmata committed:
Update metadata

check-spelling run (push) for add-subtopologies

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

moisesmata committed:
Update comment for clarity

moisesmata committed:
Initial move to phasing components

moisesmata committed:
Merge branch 'add-subtopologies' of https://github.com/moisesmata/fprime into add-subtopologies

moisesmata committed:
Fixing Phasing: Adding initial Phasing for CDHCore, DataProducts, FileHandling

moisesmata committed:
Renaming topologydefs.hpp, phasing fully working on all four subtopologies, needs cleaning up

moisesmata committed:
Cleanup with Ports_ComPacketQueue, definition only in Comms.fpp now

moisesmata committed:
CDHCore subtopology added and integrated within Ref deployment

thomas-bc commented on Svc/Subtopologies/CDHCore/SubtopologyTopologyDefs.hpp:
```diff
@@ -0,0 +1,17 @@
+
+  namespace PingEntries {
+    struct CDHCore_cmdDisp    { enum { WARN=3, FATAL=5 }; };
+    struct CDHCore_events     { enum { WARN=3, FATAL=5 }; };
```
move these ping entries to a config header

thomas-bc commented on Svc/Subtopologies/CDHCore/CDHCore.fpp:
```diff
@@ -0,0 +1,72 @@
+
+    instance fatalAdapter: Svc.AssertFatalAdapter base id CDHCoreConfig.BASE_ID + 0x0700
+
+    instance fatalHandler: Svc.FatalHandler base id CDHCoreConfig.BASE_ID + 0x0800
```
Remove from here, and add to config file

moisesmata replied:
Thoughts on just removing fatalHandler from the subtopology? I found that adding the connections instance into the config file was giving me trouble, as it was looking for the topology itself in two different files.

thomas-bc commented on Svc/Subtopologies/CDHCore/CDHCoreConfig/CDHCoreConfig.fpp:
```diff
@@ -0,0 +1,49 @@
+        phase Fpp.ToCpp.Phases.configComponents """
+        CDHCore::tlmSend.setPacketList(
+            Ref::Ref_RefPacketsTlmPackets::packetList, 
+            Ref::Ref_RefPacketsTlmPackets::omittedChannels, 
```
Need to update

thomas-bc commented on Svc/Subtopologies/CDHCore/CDHCoreConfig/CDHCoreConfig.fpp:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    module Priorities {
+        constant cmdDisp     = 101
```
priorities need to start at 99

moisesmata committed:
Merge branch 'devel' into add-cdhcore-subtopology

moisesmata committed:
Fixed to work with CCSDS, added comment on RefPackets, fatalHandler within CDHCore config

moisesmata committed:
Merge branch 'devel' into add-cdhcore-subtopology

moisesmata committed:
Add correct stackSize in CDHCoreconfig

moisesmata committed:
CDHCore Subtopology, parameterization done through config overrides
 - Subtopology with core CDH functionality within Svc
 - integrated within Ref Deployment
 - PingEntries.hpp within CDHCore/ that provides default values for PingEntries
 - Default TlmChan, option to use TlmPacketizer through changing CDHCoreConfig.fpp

moisesmata committed:
Fixed priority values in CDHCoreConfig, removed FatalHandler from CDHCore

moisesmata committed:
Update metadata

check-spelling run (pull_request_target) for add-cdhcore-subtopology

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

moisesmata committed:
Added CDHCore as a dependency for TestDeployment, in CMakeLists.txt

thomas-bc commented on Ref/Top/RefTopologyDefs.hpp:
```diff
@@ -51,49 +52,25 @@ struct TopologyState {
+    U16 port;
+};
+
+namespace PingEntries = ::PingEntries;
```
Why not define the namespace inside `namespace Ref {}` directly ? Is it needed in both places?

moisesmata replied:
I believe PingEntries cannot be inside of Ref because the subtopologies also add things to this global PingEntries namespace, and we don't want to reference the deployment from the subtopology. 
I just got a version to work where the PingEntries namespace  and the subtopology ping entries include is within namespace as well. Is that preferrable?

moisesmata replied:
Added this to the new commit

thomas-bc commented on cmake/test/data/TestDeployment/CMakeLists.txt:
```diff
@@ -23,3 +23,4 @@ add_fprime_subdirectory(./TestRelative)
 set(SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/Main.cpp")
 set(MOD_DEPS  Svc_CmdDispatcher TestLibrary_TestComponent TestLibrary2_TestComponent)
 register_fprime_deployment()
+add_dependencies(TestDeployment Svc_Subtopologies_CDHCore)
```
Does adding it to `MOD_DEPS` not work ?

moisesmata replied:
Michael helped me through this issue a bit last week, adding to `MOD_DEPS` does not properly generate the dictionary for the subtopology, so we found that doing the normal cmake `add_dependencies` was a bandaid fix, but a permanent fix will be needed in the future.

thomas-bc commented on Ref/Top/RefTopology.cpp:
```diff
@@ -96,8 +97,8 @@ void configureTopology() {
     // Health is supplied a set of ping entires.
-    health.setPingEntries(ConfigObjects::Ref_health::pingEntries,
-                          FW_NUM_ARRAY_ELEMENTS(ConfigObjects::Ref_health::pingEntries), HEALTH_WATCHDOG_CODE);
+    //CDHCore::health.setPingEntries(ConfigObjects::CDHCore_health::pingEntries,
```
Remove commented out code here and in other places

thomas-bc commented on Svc/Subtopologies/CDHCore/PingEntries.hpp:
```diff
@@ -0,0 +1,10 @@
+    struct CDHCore_tlmSend    { enum { WARN=3, FATAL=5 }; };
+  }
+
+#endif
```
Could be nice to fix those, but not required

thomas-bc commented on Ref/Top/instances.fpp:
```diff
@@ -33,155 +33,127 @@ module Ref {
 
   @ Communications driver. May be swapped with other comm drivers like UART
-  instance comDriver: Drv.TcpClient base id 0x4000
+  instance comDriver: Drv.TcpClient base id 0x1700
```
I believe we used to leave some room in between on purpose, to be able to add components easily. Arguable 2600 -> 4000 was too much.

moisesmata replied:
Added more room! I'll keep it as it was.

moisesmata committed:
Changed Structure in RefTopologyDefs.hpp to make simpler

moisesmata committed:
Reverted to prior RefTopologydefs.hpp Structure

thomas-bc reviewed (approved):
LGTM - nice work! 🚀

thomas-bc commented on Svc/Subtopologies/CDHCore/PingEntries.hpp:
```diff
@@ -0,0 +1,10 @@
+  namespace PingEntries {
+    struct CDHCore_cmdDisp    { enum { WARN=3, FATAL=5 }; };
+    struct CDHCore_events     { enum { WARN=3, FATAL=5 }; };
+    struct CDHCore_tlmSend    { enum { WARN=3, FATAL=5 }; };
```
No

thomas-bc commented on Ref/test/int/ref_integration_test.py:
```diff
@@ -80,9 +80,9 @@ def test_send_command(fprime_test_api):
     """
-    fprime_test_api.send_and_assert_command("Ref.cmdDisp.CMD_NO_OP", max_delay=0.1)
+    fprime_test_api.send_and_assert_command("CDHCore.cmdDisp.CMD_NO_OP", max_delay=0.1)
     assert fprime_test_api.get_command_test_history().size() == 1
-    fprime_test_api.send_and_assert_command("Ref.cmdDisp.CMD_NO_OP", max_delay=0.1)
+    fprime_test_api.send_and_assert_command("CDHCore.cmdDisp.CMD_NO_OP", max_delay=0.1)
```
No

LeStarch reviewed (approved):
LGTM
